dattri/func/fisher.py

```diff
@@ -405,146 +328,118 @@ def _estimate_lambda(
-                        ds_curr_reshaped[:, ts, :, None]
-                        @ a_prev_reshaped[:, ts, None, :]
-                    )
-
-            # An equivalent way to calculate lambda's. Please refer to
-            # https://math.stackexchange.com/questions/1879933/vector-multiplication-with-multiple-kronecker-products
-            batch_lambda = torch.square(q_s @ batch_dtheta @ q_a.T).mean(axis=0)
-
-            # Update the running eigenvalue estimation
-            if idx >= len(layer_lambda):
-                # First time initializartion
-                layer_lambda.append(batch_lambda)
-            else:
-                old_weight = total_samples / (total_samples + batch_samples)
-                new_weight = batch_samples / (total_samples + batch_samples)
-                layer_lambda[idx] = (
-                    old_weight * layer_lambda[idx] + new_weight * batch_lambda
-                )
+        # An equivalent way to calculate lambda's. Please refer to
+        # https://math.stackexchange.com/questions/1879933/vector-multiplication-with-multiple-kronecker-products
+        q_a, q_s = cached_q[layer_name]
+        batch_lambda = torch.square(q_s @ batch_dtheta @ q_a.T).mean(axis=0)
+
+        # Update the running eigenvalue estimation
+        if layer_name in curr_estimate:
+            old_weight = total_samples / (total_samples + batch_samples)
+            new_weight = batch_samples / (total_samples + batch_samples)
+            curr_estimate[layer_name] = (
+                old_weight * curr_estimate[layer_name] + new_weight * batch_lambda
+            )
+        else:
+            # First time initializartion
+            curr_estimate[layer_name] = batch_lambda
 
     return curr_estimate
 
 
-def ifvp_at_x_ekfac(
+def estimate_covariance(
     func: Callable,
-    *x,
-    in_dims: Optional[Tuple] = None,
-    batch_size: int = 1,
+    dataloader: torch.utils.data.DataLoader,
+    layer_cache: Dict[str, Tuple[torch.tensor]],
     max_iter: Optional[int] = None,
-    mlp_cache: Union[MLPCache, List[MLPCache]],
-    damping: float = 0.0,
-) -> Callable:
-    """IFVP via EK-FAC algorithm.
-
-    Standing for the inverse-FIM-vector product, returns a function that,
-    when given vectors, computes the product of inverse-FIM and vector.
-
-    EK-FAC algorithm provides layer-wise approximation for the IFVP function.
-    The samples are estimated based on Gauss-Newton Hessian.
+    device: Optional[str] = "cpu",
+) -> Dict[str, Tuple[torch.tensor]]:
+    """Estimate the 'covariance' matrices S and A in EK-FAC IFVP.
 
     Args:
         func (Callable): A Python function that takes one or more arguments.
             Must return the following,
-            - losses: a tensor of shape (batch_size,).
+            - loss: a single tensor of loss. Should be the mean loss by the
+                    batch size.
             - mask (optional): a tensor of shape (batch_size, t), where 1's
```

Comment:
maybe we need to make this mask output clear in the init function's document of EKFAC IF Attributor